File: storekit-offering-completing-and-restoring-in-app-purchases/Objective-C/Common/Utilities.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
A class that provides purchase data and creates an alert.
*/

@import StoreKit;
#import "Utilities.h"
#import "StoreObserver.h"
#import "AppConfiguration.h"

@implementation Utilities

- (instancetype)init {
    self = [super init];
    if (self != nil) {
        _restoreWasCalled = NO;
    }
    return self;
}

/// - returns: An array with the product identifiers to query.
-(NSArray *)identifiers {
    NSURL *url = [[NSBundle mainBundle] URLForResource:PCSProductIdsPlistName withExtension:PCSProductIdsPlistFileExtension];
    return [NSArray arrayWithContentsOfURL:url];
}

/// - returns: An array for populating the Purchases view.
-(NSMutableArray *)dataSourceForPurchasesUI {
    NSMutableArray *dataSource = [[NSMutableArray alloc] initWithCapacity:0];
    NSArray *purchased = [[StoreObserver sharedInstance].productsPurchased copy];
    NSArray *restored = [[StoreObserver sharedInstance].productsRestored copy];
    
    if (self.restoreWasCalled && (restored.count > 0) && (purchased.count > 0)) {
        dataSource = [[NSMutableArray alloc] initWithObjects:[[Section alloc] initWithName:PCSPurchasesPurchased elements:purchased],
                      [[Section alloc] initWithName:PCSPurchasesRestored elements:restored], nil];
        
    } else if (self.restoreWasCalled && (restored.count > 0)) {
        dataSource = [[NSMutableArray alloc] initWithObjects:[[Section alloc] initWithName:PCSPurchasesRestored elements:restored], nil];
    } else if (purchased.count > 0) {
        dataSource = [[NSMutableArray alloc] initWithObjects:[[Section alloc] initWithName:PCSPurchasesPurchased elements:purchased], nil];
    }
    
    /*
     Display restored products only when the user taps the Restore button (iOS), Store > Restore (macOS), or Restore all restorable purchases (tvOS)
     and there are restored products.
    */
    self.restoreWasCalled = NO;
    return dataSource;
}

#pragma mark - Create Alert

#if TARGET_OS_IOS || TARGET_OS_TV
/// - returns: An alert with a given title and message.
-(UIAlertController *)alertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *defaultAction = [UIAlertAction actionWithTitle:PCSMessagesOk style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {}];
    [alert addAction:defaultAction];
    return alert;
}
#endif

#pragma mark - Parse Section of Data

/// - returns: A Section object matching the specified name in the data array.
-(Section *)parse:(NSArray *)data forName:(NSString *)name {
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"name == %@", name];
    return (Section *)([data filteredArrayUsingPredicate:predicate].firstObject);
}

@end

File: storekit-offering-completing-and-restoring-in-app-purchases/Swift/Common/Utilities.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
A class that provides purchase data and creates an alert.
*/

import StoreKit
#if os (macOS)
import Cocoa
#else
import UIKit
#endif

class Utilities {
    
    // MARK: - Properties
    
    /// Indicates whether the user initiates a restore.
    var restoreWasCalled: Bool
    
    /// - returns: An array for populating the Purchases view.
    var dataSourceForPurchasesUI: [Section] {
        var dataSource = [Section]()
        let purchased = StoreObserver.shared.purchased
        let restored = StoreObserver.shared.restored
        
        if restoreWasCalled && (!restored.isEmpty) && (!purchased.isEmpty) {
            dataSource.append(Section(type: .purchased, elements: purchased))
            dataSource.append(Section(type: .restored, elements: restored))
        } else if restoreWasCalled && (!restored.isEmpty) {
            dataSource.append(Section(type: .restored, elements: restored))
        } else if !purchased.isEmpty {
            dataSource.append(Section(type: .purchased, elements: purchased))
        }
        
        /*
           Display restored products only when the user taps the Restore button (iOS), or chooses Store > Restore (macOS) or "Restore all restorable
           purchases” (tvOS), and there are restored products.
        */
        restoreWasCalled = false
        return dataSource
    }
    
    // MARK: - Initialization
    
    init() {
        restoreWasCalled = false
    }
    
    // MARK: - Create Alert
    
    #if os (iOS) || os (tvOS)
    /// - returns: An alert with a specified title and message.
    func alert(_ title: String, message: String) -> UIAlertController {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let action = UIAlertAction(title: NSLocalizedString(Messages.okButton, comment: Messages.emptyString),
                                   style: .default, handler: nil)
        alertController.addAction(action)
        return alertController
    }
    #endif
}
